pkg/storage/etcd3/watcher.go

```diff
@@ -182,6 +194,10 @@ func (wc *watchChan) startWatching() {
 			wc.sendEvent(parseEvent(e))
 		}
 	}
+	// When we come to this point, it's possible that:
+	// - client side ends the watch, e.g. cancel the context, close the client.
+	// - server side ends the watch.
+	close(watchClosedCh)
```

Comment:
It's incomprehensible to me why you'd want to only notify your caller under that condition. It seems the caller should be notified when this function exits for any reason. Please heavily comment or change. It's really confusing right now.